// Copyright 2015 The gRPC Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.carapi.queryEngine";
option java_outer_classname = "queryEngineProto";
option objc_class_prefix = "RTG";

package queryEngine;

// Interface exported by the server.
service QueryEngine {
  // A simple RPC.
  //
  // Obtains the query at a given position.
  //
  // A feature with an empty name is returned if there's no feature at the given
  // position.
  rpc sendQuery(QueryRequest) returns (QueryResponse) {}
}


/* A query is represented as a string containing a set of wikipedia headings
 * Either in the format enwiki:/chocolate/labelling
 * Or chocolate labelling
 * A query also includes the number of results to return for that query
 * Benchmark value is 1000 */
message QueryRequest {
  string rawQueryText = 1;
  string originalQuery = 2;
  int32 numberOfDocs = 3;
}

// Query response, contains a paragraph, similarity score, and the evaluation metrics.
// Evaluation metrics are mean average precision, r-precision and non distributed cumulative gain
message QueryResponse {
  repeated result Results = 1;
  eval queryEval = 2;
  repeated string qrelParagraph = 3;
}

message result {
  string paragraph = 1;
  string similarity = 2;
}

message eval {
  float map = 3;
  float rprec = 4;
  float ndcg = 5;
}